### PR TITLE
dfp: adding an experimental option for E-M

### PR DESCRIPTION
**Pull request by alyssawilk**
This adds an off-by-default flag that allows the DFP filter to force resolution if there are no upstream connections. This doesn't work for strict DNS, and is largely intended for use for Envoy mobile.  As the TODO states, if the performance is net positive we'll add configuration knobs, otherwise it'll be torn out.

Risk Level: low (off by default)
Testing: new e2e tests
Docs Changes: n/a
Release Notes: n/a
[Optional Runtime guard:] yes

### CODE REVIEW

File: source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc [hunk 1]
```diff
@@ -239,15 +239,24 @@ Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::RequestHeaderMap& hea
...
+        dynamic_cast<Extensions::Common::DynamicForwardProxy::DfpLb*>(&cluster->loadBalancer());
+    if (dfp_lb) {
+      std::string hostname = Common::DynamicForwardProxy::DnsHostInfo::normalizeHostForDfp(
+          headers.getHostValue(), default_port);
+      auto host = dfp_lb->findHostByName(hostname);
+      if (host && !host->used()) {
+        force_cache_refresh = true;
+      }
+    }
+  }
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc:254]: I'm sure I'm missing all the context, but it's not immediately obvious to me what case this is trying to fix with the force refresh? Can you add a small comment trail about why this is being added/tested?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with a small comment request, thanks.

/wait

**Comment by alyssawilk** [comment]
OK ci sorted, PTAL

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**